nav2_util/src/CMakeLists.txt

```diff
@@ -10,34 +10,41 @@ add_library(${library_name} SHARED
   odometry_utils.cpp
   array_parser.cpp
 )
-
-ament_target_dependencies(${library_name}
-  rclcpp
-  nav2_msgs
-  tf2
-  tf2_ros
-  nav_msgs
-  geometry_msgs
-  lifecycle_msgs
-  rclcpp_lifecycle
-  tf2_geometry_msgs
-  bondcpp
+target_include_directories(${library_name}
+  PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/include>"
+  "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>")
+target_link_libraries(${library_name} PUBLIC
+  bondcpp::bondcpp
+  ${geometry_msgs_TARGETS}
+  ${lifecycle_msgs_TARGETS}
+  ${nav2_msgs_TARGETS}
+  ${nav_msgs_TARGETS}
+  ${rcl_interfaces_TARGETS}
+  rclcpp::rclcpp
+  rclcpp_action::rclcpp_action
+  rclcpp_lifecycle::rclcpp_lifecycle
+  tf2_ros::tf2_ros
+  tf2::tf2
+  ${tf2_geometry_msgs_TARGETS}
+)
+target_link_libraries(${library_name} PRIVATE
+  ${bond_TARGETS}
```

Comment:
Question: why is this one & rclcpp below private? When do we do public vs private?

Reply:
Yeah, good question.

So the difference between PUBLIC and PRIVATE is that PUBLIC dependencies are exported to downstream consumers, while PRIVATE ones are not.  Marking dependencies as PRIVATE (where they can be) is part of what helps speed up compilation, since the compiler doesn't have to go searching through additional directories for headers and libraries.

As for when to use each, there are separate rules for executables and libraries:
1.  No downstream can ever compile or link against an executable, so all dependencies can be PRIVATE.  This honestly doesn't matter too much, as no downstream will even attempt to link, but making it PRIVATE is just cleaner.
2. For libraries, a dependency is PRIVATE if it is only used during compilation of *this* package.  So in a standard ROS package layout, if something is only ever referenced in the `src` directory, then it can be PRIVATE.  Anything that is referenced in the `include` directory needs to be PUBLIC, because in order for a downstream package to compile, it also needs to have access to those includes.

As a concrete example for 2, in `nav2_util` the `bond` package (which is really a message package, despite the name) is only ever referenced in https://github.com/ros-navigation/navigation2/blob/03ee50a629425bdd2f24463474732b7d9bf6fe95/nav2_util/src/lifecycle_node.cpp , so it can be PRIVATE.  On the other hand, `geometry_msgs` is referenced in https://github.com/ros-navigation/navigation2/blob/03ee50a629425bdd2f24463474732b7d9bf6fe95/nav2_util/include/nav2_util/geometry_utils.hpp (among other places), so it needs to be PUBLIC (a library that depends on `nav2_util` needs to know where to find the `geometry_msgs` headers so it can compile).

Does this make sense?

Reply:
> * So what if the compiled library is used in another package (i.e. lifecycle node is then used in other packages as the base of the action servers)? Seem unintuitive to me that the include in the cpp vs hpp would make that much of a difference, but 👍

I'll preface this by saying I don't completely know how this works, so the below is a bit of speculation.

Thinking about it, if you have a symbol that is only ever internal to a library, then no downstream can ever make (direct) reference to it.  Thus, the downstream packages don't have to know how to resolve that upstream symbol in order to compile/link.  They *do* have to know how to resolve that symbol at runtime, but I believe that is handled either via RPATH/RUNPATH, or via LD_LIBRARY_PATH.

> But if that was in the `lifecycle_node.hpp` file, then it would need to be public, correct?

Yes, exactly.

> So.. is it worth me at some point doing a review of the stack and moving as much to cpp files to make the dependencies private?

Moving as much as possible into the `.cpp` files will reduce the number of dependencies you need to export to downstream, so yes, I think it is worth doing.